src/providers/mssql/qgsmssqlprovider.cpp

```diff
@@ -142,16 +144,34 @@ QgsMssqlProvider::QgsMssqlProvider( const QString &uri, const ProviderOptions &o
       loadMetadata();
     }
     loadFields();
-    UpdateStatistics( mUseEstimatedMetadata );
 
-    QString primaryKey = anUri.keyColumn();
-    if ( !primaryKey.isEmpty() )
+    UpdateStatistics( mUseEstimatedMetadata, mError );
+
+    //only for views, defined in layer data when loading layer for first time
+    bool primaryKeyFromGeometryColumnsTable = anUri.hasParam( QStringLiteral( "primaryKeyInGeometryColumns" ) )
+        ? anUri.param( QStringLiteral( "primaryKeyInGeometryColumns" ) ).toInt()
+        : false;
+
+    QStringList cols;
+    if ( primaryKeyFromGeometryColumnsTable )
```

Comment:
shall we add a condition that this is only for views - just to make it clear?

Reply:
here we don't if this is a view or not, but maybe I miss something...